### PR TITLE
^-input pattern to exclude a specific file from the previously listed

### PR DESCRIPTION
**Pull request by naudhr**
Adding ability to exclude some files from, for example, «:foreach» rule.

Useful in the case when all the library sources are placed into one directory and just a few of them are used to build some special tool (like script generator, testtapp, supervisor app, etc) with special dependencies. So one can «:foreach *.c ^main.c ^test.c», and then add specific rules for excepted modules.

### CODE REVIEW

**Comment by hidroto** [comment]
would this work with a directory with files 
- a.file
- b.file
- c.file

and Tupfile

: *.file ^b.file |> wc %f > %o | lines_in_a_and_c.file

to make a file lines_in_a_and_c.file with word count of a.file and c.file without b.file.

**Comment by naudhr** [comment]
Certainly. That is what this patch was written for :)
The test suggested demonstrates this behaviour, though now I see it looks a bit clumsy comparing to your case. Should I add your example in my test?

**Comment by hidroto** [comment]
i think another improvement would be to have negated globs instead of just plan strings.
like this
ls 
a.file b.file c.file d.file

> : *.file ^[bc].file |> wc %f > %o |> lines_in_a_and_d.file

but i admit that looks to be much harder to achieve.

**Comment by naudhr** [comment]
Just added negated globs support.
The syntax is identical to the ordinary globs, just preceded by a «^».

File: src/tup/parser.c [hunk 1]
```diff
@@ -2732,9 +2739,11 @@ static int nl_add_path(struct tupfile *tf, struct path_list *pl,
 		struct variant *variant;
 
 		if(tup_db_select_tent_part(pl->dt, pl->pel->path, pl->pel->len, &tent) < 0) {
-			return -1;
+			return /*args.excluding ? 0 :*/ -1;
```
Comment (gittup) [inline src/tup/parser.c]: Did you still need to do something with args.excluding here? I'm not sure why this is commented out.

File: src/tup/parser.c [hunk 2]
```diff
@@ -2732,9 +2739,11 @@ static int nl_add_path(struct tupfile *tf, struct path_list *pl,
...
 		}
 		if(!tent || tent->type == TUP_NODE_GHOST) {
+			//if(args.excluding)
+			//	return 0;
```
Comment (gittup) [inline src/tup/parser.c]: Similarly here.

File: src/tup/parser.c [hunk 3]
```diff
@@ -2763,12 +2772,12 @@ static int nl_add_path(struct tupfile *tf, struct path_list *pl,
 		variant = tup_entry_variant(tent);
 		if(!variant->root_variant && variant != tf->variant) {
 			fprintf(tf->f, "tup error: Unable to use files from another variant (%s) in this variant (%s)\n", variant->variant_dir, tf->variant->variant_dir);
-			return -1;
+			return /*args.excluding ? 0 :*/ -1;
```
Comment (gittup) [inline src/tup/parser.c]: And here :)

File: test/input-excheptions.sh [hunk 4]
```diff
@@ -0,0 +1,36 @@
+#! /bin/sh -e
```
Comment (gittup) [inline test/input-excheptions.sh]: How come this test exists both as input-excheptions.sh and t3990-input-exceptions.sh?

**Comment by gittup** [comment]
Overall this looks very good! Mostly I just wasn't sure what to do about the commented out code, otherwise it looks about ready to merge.

**Comment by naudhr** [comment]
Oh, how red-faced I am...
For sure, I've had to clean everything up and then double-check it.

The `input-excheptions.sh` was the first variant of the test, which i've given up in favour of the `wc`.

File: test/tup.sh [hunk 5]
```diff
@@ -247,10 +247,10 @@ __update()
 	fi
 
 	set +e
-	if [ -z $@ ]; then
-		$cmd "$@"
+	if [ $# -le 0 ]; then
```
Comment (gittup) [inline test/tup.sh]: What are the changes to this file supposed to fix? All tests passed without them on my machines. Is this a separate issue that we need to address for Gentoo?

**Comment by gittup** [comment]
I've merged most of your commits. I had to make a few small changes to get the test working on all platforms ('wc' has a different output format on OSX, so I added content to the files and grepped for that instead). Also I didn't merge the test/tup.sh changes since I'm not sure what they're for.

**Comment by hidroto** [comment]
i have triade to get this feature to work for some of my builds but i am having a problum with subdirs not being exclude at all

this test case should succed but does not

    . ./tup.sh

    echo ': sub/*.file sub/^c.file |> cat %f > %o |> lines.txt' > Tupfile

    mkdir sub
    cd sub
    for i in a b c d; do
    	echo $i > $i.file
    done
    cd ..
    update

    gitignore_good a lines.txt
    gitignore_good b lines.txt
    gitignore_bad c lines.txt
    gitignore_good d lines.txt

am i right in how this feature should behave or am i using the testscript wrong?

**Comment by naudhr** [comment]
Hmmm, firstly, the rule is supposed to be like this: `: sub/*.file ^sub/c.file |> ... anything`, where `^` emerges as the first symbol of the input token and thus negates the globbed rest of the token. I had to spend some time on the description of the feature.

Anyway, it seems that the subdir doesn't work, failing with `Failed to find directory ID for dir '^sub/c.file' relative to '.'`. I will look into.

**Comment by naudhr** [comment]
@gittup : the `[ -z ]` expression on the semi-self-hacked-slackware-box of the colleague of mine brings `sh: [: -z: unary operator expected`. Deadly mysterious, but checking the arg count instead of the arg string emptiness solved the issue.
I'm working on pulling more and more people into my tup-band so trying to get rid of any error message.

**Comment by gittup** [comment]
Ok, I've added the test script changes as well. Feel free to file a separate issue for such things so it's easy to see what changes fix the different issues. Thanks!

**Comment by naudhr** [comment]
Fixed excluding in subdirs.
Also added a testcase that excludes an absent file.

**Comment by gittup** [comment]
Added the subdir fix. Thanks a bunch!

**Comment by naudhr** [comment]
Merged in.

**Pull request closed without merging**